### PR TITLE
Ia32 structs

### PR DESCRIPTION
**Pull request by P-E-P**
Flatten structs and use more meaningful names.

### CODE REVIEW

File: src/arch/ia32/descriptor/gate/interrupt/configuration.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /// Offset of the present bit (P) within the
+    /// [Configuration](super::Configuration) structure.
+    pub const PRESENT: usize = 15;
+}
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0600)
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/gate/interrupt/configuration.rs:43]: Maybe a comment describing in details what the default configuration is, could help

File: src/arch/ia32/descriptor/gate/task/configuration.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /// Offset of the present bit (P) within the
+    /// [Configuration](super::Configuration) structure.
+    pub const PRESENT: usize = 15;
+}
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0500)
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/gate/task/configuration.rs:35]: Here too, we could give details

File: src/arch/ia32/descriptor/gate/trap.rs [hunk 3]
```diff
@@ -1,41 +1,91 @@
...
-                .segment_selector(segment_selector.into()),
-            upper: Upper::default().offset_high(offset_31_16).present(1),
+            offset_15_0: offset.get_bits(0..16).try_into().unwrap(),
+            offset_31_16: offset.get_bits(16..32).try_into().unwrap(),
+            configuration: Configuration::default().present(true),
+            segment_selector,
         }
     }
 
+    /// Change a [``]'s privilege level by another one.
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/gate/trap.rs]: It think the `TrapGate` word is missing between the brackets.

File: src/arch/ia32/descriptor/gate/trap/configuration.rs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    }
+
+    /// Offset of the present bit (P) within the
+    /// [Configuration](super::Configuration) structure.
+    pub const PRESENT: usize = 15;
+}
+
+impl Default for Configuration {
+    fn default() -> Self {
+        Configuration(0x0700)
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/gate/trap/configuration.rs:42]: Here again, if you agree with specifying the bits set

File: src/arch/ia32/descriptor/gate/trap/configuration.rs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+        Self(*self.0.clone().set_bit(PRESENT, present))
+    }
+}
+
+impl fmt::Display for Configuration {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use offset::{privilege_level, PRESENT, SIZE};
+        write!(
+            f,
+            "Size: {}\nDPL: {}\nPresent {}",
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/gate/trap/configuration.rs]: Here, and in every other `Configuration` display traits, the semicolon `:`  is missing in after "Present"

File: src/arch/ia32/descriptor/segment.rs [hunk 6]
```diff
@@ -1,155 +1,301 @@
...
+use permissions::Permissions;
 
-pub mod lower;
-mod upper;
+mod configuration;
+mod permissions;
 
+/// The default operation size modes of a memory segment.
+///
+/// It could be either 16 bits (`0`) or 32 bits (`0`). It performs different
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/segment.rs]: I believe it was supposed to be `or 32 bits ('1')`

File: src/arch/ia32/descriptor/segment.rs [hunk 7]
```diff
@@ -1,155 +1,301 @@
...
+/// functions depending on wheter the segment descriptor is an executable code
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/segment.rs]: `whether`

File: src/arch/ia32/descriptor/tss/permissions.rs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    /// Offset of the present bit (P) within the
+    /// [Permissions](super::Permissions) structure.
+    pub const PRESENT: usize = 7;
+}
+
+impl Default for Permissions {
+    fn default() -> Self {
+        Permissions(0b00001001)
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/tss/permissions.rs:43]: Here again, do you want to precise what is set ?

File: src/arch/ia32/descriptor/tss/permissions.rs [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        Self(*self.0.clone().set_bit(PRESENT, present))
+    }
+}
+
+impl fmt::Display for Permissions {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use offset::{privilege_level, BUSY, PRESENT};
+        write!(
+            f,
+            "Busy: {}\nDPL: {}\nPresent {}",
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/tss/permissions.rs]: Here too if you want to put a semicolon after "Present"

File: src/arch/ia32e/descriptor/gate.rs [hunk 10]
```diff
@@ -1,92 +1,83 @@
...
 #[must_use]
 #[derive(Default, Copy, Clone)]
 #[repr(C, packed)]
 pub struct Gate {
-    /// Segment selector bits (0:31).
-    lower: Lower,
-    ///  Segment selector bits (32:64).
-    upper: Upper,
-    /// Procedure's entry point offset high bits.
+    offset_15_0: u16,
```
Comment (alexandrefresnais) [inline src/arch/ia32e/descriptor/gate.rs:31]: Don't you want to comment the role of every field as you did for other structs ?

File: src/arch/ia32/descriptor/tss.rs [hunk 11]
```diff
@@ -1,65 +1,139 @@
...
+mod configuration;
+mod permissions;
 
+/// A task state segment descriptor structure that can be used directly by the
+/// processor to describe a task state segment.
 #[must_use]
-#[derive(Copy, Clone)]
+#[derive(Default, Copy, Clone)]
 #[repr(C, packed)]
 pub struct TssDescriptor {
```
Comment (alexandrefresnais) [inline src/arch/ia32/descriptor/tss.rs:18]: Don't you want to implement `fmt::Display` for this structure ?
You can leave a `todo!` if you want as in the ia32e gate.rs.

File: src/arch/ia32e/descriptor/tss.rs [hunk 12]
```diff
@@ -1,59 +1,123 @@
...
 #[repr(C, packed)]
 pub struct TssDescriptor {
     tss: IA32TssDescriptor,
+    /// Bits 32 to 63 of the segment base address.
     base_63_32: u32,
+    /// Reserved bits.
     reserved: u32,
 }
 
 impl TssDescriptor {
```
Comment (alexandrefresnais) [inline src/arch/ia32e/descriptor/tss.rs:21]: No `fmt::display` implementation here too ?

**Review by alexandrefresnais — CHANGES_REQUESTED** [review]
Mad impressive.
I left some minor comments.
Sorry and thank you for correcting the coding style in `address.rs`.

File: src/arch/ia32/descriptor/gate/interrupt/configuration.rs (hunk 1 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/gate/interrupt/configuration.rs:43]: Done in ba5c2c8

File: src/arch/ia32/descriptor/gate/task/configuration.rs (hunk 2 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/gate/task/configuration.rs:35]: Done in ba5c2c8

File: src/arch/ia32/descriptor/gate/trap/configuration.rs (hunk 4 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/gate/trap/configuration.rs:42]: Done in ba5c2c8

File: src/arch/ia32/descriptor/gate/trap.rs (hunk 3 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/gate/trap.rs]: Done in 801000d

File: src/arch/ia32/descriptor/tss/permissions.rs (hunk 9 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/tss/permissions.rs]: Done in 57418d7

File: src/arch/ia32/descriptor/gate/trap/configuration.rs (hunk 5 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/gate/trap/configuration.rs]: Done in 57418d7

File: src/arch/ia32/descriptor/tss/permissions.rs (hunk 8 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/tss/permissions.rs:43]: Done in ce93a17

File: src/arch/ia32/descriptor/tss.rs (hunk 11 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/tss.rs:18]: The current display format is not the final one, I'm not too sure about what we should do or what we should need. We'll keep the current code and implement proper display trait for those  structure in another PR.

File: src/arch/ia32e/descriptor/tss.rs (hunk 12 above)
Comment (P-E-P) [reply src/arch/ia32e/descriptor/tss.rs:21]: The current display format is not the final one, I'm not too sure about what we should do or what we should need. We'll keep the current code and implement proper display trait for those  structure in another PR.

File: src/arch/ia32/descriptor/segment.rs (hunk 6 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/segment.rs]: Fixed in 141a5a9

File: src/arch/ia32/descriptor/segment.rs (hunk 7 above)
Comment (P-E-P) [reply src/arch/ia32/descriptor/segment.rs]: Fixed in 141a5a9

File: src/arch/ia32e/descriptor/gate.rs (hunk 10 above)
Comment (P-E-P) [reply src/arch/ia32e/descriptor/gate.rs:31]: Fixed in fe6ea60

**Review by alexandrefresnais — APPROVED** [review]
LGTM, very well done!

**Pull request merged**